Add support for raw HTML

flip111 opened the pull request:
tests pass
most examples are working. The ones that don't i don't think it's because of this PR but didn't look into it
comes with example of usage

Addresses https://github.com/purescript-halogen/purescript-halogen/issues/324

* Not sure for the best way to implement patch
* [ComponentSlot](https://github.com/flip111/purescript-halogen/blob/raw-html-widget/src/Halogen/Component.purs#L204) should be renamed to `HalogenWidget` imo because the name doesn't cover the extra constructor. This would be a breaking change, while currently what is added i don't think it breaks anything so it could be updated in minor version of halogen library.
* add documentation, only first element in raw html string will be used
* rename function `rawHTML` just to `html` or `raw`?

Review appreciated.

flip111 committed:
Add support for raw HTML

MonoidMusician commented on src/Halogen/VDom/Driver.js:
```diff
@@ -0,0 +1,12 @@
+  var template = document.createElement('template');
+  html = html.trim(); // Never return a text node of whitespace as the result
+  template.innerHTML = html;
+  var newElement = template.content.firstChild;
```
Does this mean it only supports a single HTML element? That should at least be documented somewhere. Maybe with support for fragments it could be improved to capture all children.

flip111 replied:
I hadn't considered that .. but good point. I guess it does mean that.

It doesn't necessarily need to be like this but inserting multiple elements would need to happen at the parent of the current element. Which i think would make the code unnecessary more complicated and might not even be possible (what to do if the raw-html was the top halogen element).

Also i think it's unwanted to enter multiple nodes in one location. Because you can just put a node around it. i.e. `<b>node1</b><b>node2</b>` into `<div><b>node1</b><b>node2</b></div>`.

What could be done is count the amount of children and if it's more than 1 to show a [console warning](https://developer.mozilla.org/en-US/docs/Web/API/Console/warn) on runtime.

Yes and documentation.

flip111 committed:
Merge remote-tracking branch 'upstream/master' into raw-html-widget

flip111 committed:
Update to purescript 0.15

* Console error on passing not exactly 1 element to rawHTML function

flip111 commented:
[How can i compare the html string](https://github.com/flip111/purescript-halogen/blob/raw-html-widget/src/Halogen/VDom/Driver.purs#L109)

How can i make this more efficient ? Like comparing the new html string with the old one and not render in case they are the same ?

flip111 committed:
Better implementation for RawHTML patch function

flip111 committed:
Merge remote-tracking branch 'upstream/master' into raw-html-widget

flip111 committed:
Fix RawHTML widget DOM lifecycle

The original implementation used Nothing as the widget state for
RawHTML nodes. This meant the patch function couldn't access the old
DOM node to remove it, and the halt function couldn't clean up.
Every re-render created a new element via insertChildIx without
removing the old one, causing duplicate nodes to accumulate.

Fix by replacing the Maybe-based WidgetState with a proper sum type
(ThunkWidget | RawHTMLWidget | NoWidget) so each widget variant
tracks the state it needs for correct patching and cleanup.

Also fixes JS FFI bugs:
- outerHtml -> outerHTML (wrong casing, returned undefined)
- string concat . -> + (PHP syntax, not JS)
- console.error -> throw Error (fail loud on invalid input)
- Replace outerHTML setter approach with replaceWith() to avoid
  stale node references

Merge upstream/master (6 commits, no conflicts).
Expand raw-html example to cover dynamic, conditional, and mixed
content scenarios.
